doc/addrinfo.1
.\" ***********************************************************************
.\" *  Copyright © 2006 Rémi Denis-Courmont.                              *
.\" *  This program is free software; you can redistribute and/or modify  *
.\" *  it under the terms of the GNU General Public License as published  *
.\" *  by the Free Software Foundation; version 2 of the license.         *
.\" *                                                                     *
.\" *  This program is distributed in the hope that it will be useful,    *
.\" *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
.\" *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               *
.\" *  See the GNU General Public License for more details.               *
.\" *                                                                     *
.\" *  You should have received a copy of the GNU General Public License  *
.\" *  along with this program; if not, you can get it from:              *
.\" *  http://www.gnu.org/copyleft/gpl.html                               *
.\" ***********************************************************************
.TH "ADDRINFO" "1" "$Date$" "addrinfo" "User's Manual"
.SH NAME
addrinfo, nameinfo \- perform DNS lookups from scripts
.SH SYNOPSIS
.BR "addrinfo" " [" "-46mnr" "] [" "HOSTNAMES" "]"

.BR "nameinfo" " [" "-46n" "] [" "HOSTNAMES" "]"

.SH DESCRIPTON
Write addresses or canonical hostname of specified names to the standard
output. If no names are specified via the command line, they are read from
the standard input.

.SH OPTIONS

.TP
.BR "\-4" " or " "\-\-ipv4"
Only try to perform IPv4 lookups.

.TP
.BR "\-6" " or " "\-\-ipv6"
Only try to perform IPv6 lookups.

.TP
.BR "\-h" " or " "\-\-help"
Display some help and exit.

.TP
.BR "\-m" " or " "\-\-multiple"
Normally, if a name yields multiple results, only the first one is printed.
With this optional parameter, all results will be printed on a single line,
separated by spaces.

.TP
.BR "\-n" " or " "\-\-numeric"
Prevent forward hostname lookup.
This ensures that the input names are numeric addresses, which do not need
to be looked up. Typically used along with the \-\-reverse option to convert
numerical addresses to canonical hostnames.

.TP
.BR "\-r" " or " "\-\-reverse"
Perform a reverse DNS lookup
.RB "(enabled implicitly with " "addrinfo" ")."
.BR "addrinfo" " prints numerical resolved addresses by default instead."

.TP
.BR "\-\-version"
Display program version and license and exit.

.SH "SEE ALSO"
host(1), getaddrinfo(3), getnameinfo(3), resolv.conf(5)

.SH AUTHOR
R\[char233]mi Denis-Courmont <rdenis at simphalempin.com>

$Id$

http://www.remlab.net/ndisc6/


NEWS
RELEASE NOTES for NDisc6
version 0.9.0
=========================
$Id$

  There are no known issues at the moment.
Please report any problem by email at address :
rdenis (at) simphalempin (dot) com
Thanks for your help.

  You can find major changes since version 0.1.5 below. For more general
informations on this program, see README. For syntax details, read the
provided manual pages.

============================================================================
TEST RELEASE 0.9.0 : Major features enhancement, major bug fixes

# rdisc6 supports RFC4191 (router preference and specific routes)
# rdisc6 uses RFC2461 probe interval by default (4" instead of 1"). 
# traceroute6 parallelized in a safe manner.
# traceroute6 handles Parameter Problem / Next Header errors better.
# traceroute6 prints numerical IPv6 address when DNS fails.
# traceroute6 includes IPv6 and extensions headers in packet size.
# tcpspray: fix rare error handling deadlock.

============================================================================
STABLE RELEASE 0.7.4 : Minor feature enhancement

# traceroute6 can send TCP probes of different sizes.

============================================================================
STABLE RELEASE 0.7.3 : Minor features enhancement, major bug fixes

# traceroute6 won't stop at the first non-responsive hop anymore.
# traceroute6 can print hop limit of received packet with option -l.
# Support for resolving IDN (Internationalized Domain Names).

============================================================================
STABLE RELEASE 0.7.2 : Minor features enhancement

# rdisc6, ndisc6 and traceroute6 now drop root privilege at the very
  beginning of execution to prevent all kind of would-be local root
  compromise (in case of bug in the command line parsing code).
# FreeBSD build fixes.

============================================================================
STABLE RELEASE 0.7.1 : Minor portability fixes

# Traffic class and routing header compatibility code for old libc
  (on Linux, FreeBSD and NetBSD).
# Fix clock_nanosleep availability detection.

============================================================================
STABLE RELEASE 0.7.0 : Major feature enhancement

# nameinfo and addrinfo DNS resolver for scripting added.
# dnssort, simple sort script for DNS hostnames added.
# IPv6 "Type 0" (loose source) routing support for traceroute6.
# Outgoing interface selection with traceroute6 (even on non-Linux).
# Minor fixes for LP64 architectures (incl. AMD64).
# traceroute6 can skip IPv6 extension headers.

============================================================================
STABLE RELEASE 0.6.8 : Major bug fixes

# Fix (very) incorrect duration and bandwidth result from tcpspray6.
# Support for specifying IPv6 Traffic Class with traceroute6.

============================================================================
STABLE RELEASE 0.6.7 : Minor security fix

# Check hop limit of incoming neighbor discovery packets (ndisc6, rdisc6).

============================================================================
STABLE RELEASE 0.6.6 : Minor portability fixes

# FreeBSD compilation fix.

=============================================================================
STABLE RELEASE 0.6.5 : Minor feature enhancement

# Use POSIX monotonic clock instead of the real-time clock

=============================================================================
STABLE RELEASE 0.6.4 : Minor bug fixes

# Autoconf updates.
# Fix non-ASCII characters within manual pages.
# Debian packaging support.

=============================================================================
STABLE RELEASE 0.6.3 : Minor bug fixes

# tcptraceroute6 handles parameters with white spaces correctly.
# Renamed traceroute6 to rltraceroute6 to avoid conflict with iputils.

=============================================================================
STABLE RELEASE 0.6.2 : Major feature enhancement, minor security fixes

# Support for ICMPv6 Echo requests traceroute6.
# traceroute6 support for specifying UDP and ICMPv6 packets byte size
  (payload no longer has an empty payload by default).
# Support for sending ECN-setup SYN packets with tcptraceroute6.
# Support for selecting base UDP destination port with traceroute6.
# Support for selecting source port number with tcptraceroute6.
# Support for custom delay in-between each (tcp)traceroute6 probes.
# Sanitize raw socket file descriptors number before use.
# Fix symlink installation wrt DESTDIR (such as when packaging via RPM).

=============================================================================
STABLE RELEASE 0.6.0 : Major feature enhancement, portability fixes

# Switch to the bloated autotools.
# rdisc6 and traceroute6 compile fixes for Solaris.
# ndisc6 fixes for BSD and Solaris.
# tcpspray6 Discard/Echo bandwidth metter added.

=============================================================================
TEST RELEASE 0.5.3 : Minor bug fixes

# Port (tcp)traceroute6 to BSD.
# Fix UDP traceroute6 regression from version 0.5.2.
# Handle ICMPv6 Parameter problem errors.

=============================================================================
TEST RELEASE 0.5.2 : Minor bug fixes

# Fix minor input buffer boundary check error.
# Fix detection of invalid hop limits on the command line.

=============================================================================
TEST RELEASE 0.5.1 : Minor feature enhancement

# Add traditional UDP traceroute to traceroute6 (formerly tcptraceroute6).
# tcptraceroute6 is now an alias for “traceroute6 -S”.
# Source address specification option for (tcp)traceroute6.
# TCP/ACK probes option for (tcp)traceroute6.
# Use a rather high source TCP port, dependant on PID.

=============================================================================
TEST RELEASE 0.5.0 : Major feature enhancement

# tcptraceroute6 was added to the package.


src/addrinfo.c
/*
 * addrinfo.c - converts names to network addresses
 * $Id$
 */

/***********************************************************************
 *  Copyright (C) 2002-2006 Rémi Denis-Courmont.                       *
 *  This program is free software; you can redistribute and/or modify  *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; version 2 of the license.         *
 *                                                                     *
 *  This program is distributed in the hope that it will be useful,    *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               *
 *  See the GNU General Public License for more details.               *
 *                                                                     *
 *  You should have received a copy of the GNU General Public License  *
 *  along with this program; if not, you can get it from:              *
 *  http://www.gnu.org/copyleft/gpl.html                               *
 ***********************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <string.h> /* strchr() */
#include <stdbool.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <unistd.h>
#include <locale.h>
#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif

#ifndef AI_IDN
# define AI_IDN 0
#endif

static void
gai_perror (int errval, const char *msg)
{
	if (errval == EAI_SYSTEM)
		perror (msg);
	else
		fprintf (stderr, "%s: %s\n", msg, gai_strerror (errval));
}


static int
printnames (const char *name, int family, int aflags, int nflags, bool single)
{
	struct addrinfo hints, *res;

	memset (&hints, 0, sizeof (hints));
	hints.ai_family = family;
	hints.ai_socktype = SOCK_DGRAM;
	hints.ai_flags = aflags;

	int check = getaddrinfo (name, NULL, &hints, &res);
	if (check)
	{
		gai_perror (check, name);
		return -1;
	}

	for (struct addrinfo *ptr = res; ptr != NULL; ptr = ptr->ai_next)
	{
		char hostname[NI_MAXHOST];

		check = getnameinfo (ptr->ai_addr, ptr->ai_addrlen,
		                     hostname, sizeof (hostname),
		                     NULL, 0, nflags);

		if (check)
			gai_perror (check, name);
		else
			fputs (hostname, stdout);

		if (single)
			break;

		if (ptr->ai_next != NULL)
			fputc (' ', stdout);
	}

	fputc ('\n', stdout);
	freeaddrinfo (res);
	return 0;
}


static int
printnamesf (FILE *input, int family, int aflags, int nflags, bool single)
{
	do
	{
		char buf[NI_MAXHOST + 1], *ptr;

		if (fgets (buf, sizeof (buf), input) != NULL)
		{
			ptr = strchr (buf, '\n');
			if (ptr != NULL)
				*ptr = '\0';

			printnames (buf, family, aflags, nflags, single);
		}
	}
	while (!ferror (input) && !feof (input));

	if (ferror (input))
	{
		perror (_("Input error"));
		return -1;
	}

	return 0;
}


static int usage (const char *path)
{
	printf (_(
"Usage: %s [-4|-6] [hostnames]\n"
"Converts names to addresses.\n"
"\n"
"  -4, --ipv4     only lookup IPv4 addresses\n"
"  -6, --ipv6     only lookup IPv6 addresses\n"
"  -h, --help     display this help and exit\n"
"  -m, --multiple print multiple results separated by spaces\n"
"  -n- --numeric  do not perform forward hostname lookup\n"
"  -r, --reverse  perform reverse address to hostname lookup\n"
"  -V, --version  display program version and exit\n"), path);
	return 0;
}


static int quick_usage (const char *path)
{
	fprintf (stderr, _("Try \"%s -h\" for more information.\n"), path);
	return 2;
}


static int version (void)
{
        printf (_(
"addrinfo %s ($Rev$)\n"
" built %s on %s\n"), VERSION, __DATE__, PACKAGE_BUILD_HOSTNAME);
        printf (_("Configured with: %s\n"), PACKAGE_CONFIGURE_INVOCATION);
        puts (_("Written by Remi Denis-Courmont\n"));

        printf (_("Copyright (C) %u-%u Remi Denis-Courmont\n"
"This is free software; see the source for copying conditions.\n"
"There is NO warranty; not even for MERCHANTABILITY or\n"
"FITNESS FOR A PARTICULAR PURPOSE.\n"), 2002, 2006);
        return 0;
}

static const struct option lopts[] =
{
	{ "ipv4",     no_argument,       NULL, '4' },
	{ "ipv6",     no_argument,       NULL, '6' },
	{ "help",     no_argument,       NULL, 'h' },
	{ "multiple", no_argument,       NULL, 'm' },
	{ "numeric",  no_argument,       NULL, 'n' },
	{ "reverse",  no_argument,       NULL, 'r' },
	{ "version",  no_argument,       NULL, 'V' },
	{ NULL,       0,                 NULL, 0   }
};

static const char sopts[] = "46hmnrV";

int main (int argc, char *argv[])
{
	setlocale (LC_CTYPE, "");

	int val, family = AF_UNSPEC, aflags = AI_IDN, nflags = NI_NUMERICHOST;
	bool single = true;

	while ((val = getopt_long (argc, argv, sopts, lopts, NULL)) != EOF)
		switch (val)
		{
			case '4':
				family = AF_INET;
				break;

			case '6':
				family = AF_INET6;
				break;

			case 'h':
				return usage (argv[0]);

			case 'm':
				single = false;
				break;

			case 'n':
				aflags = AI_NUMERICHOST;
				break;

			case 'r':
				nflags &= ~NI_NUMERICHOST;
				break;

			case 'V':
				return version ();

			case '?':
			default:
				return quick_usage (argv[0]);
		}

	setvbuf (stdout, NULL, _IOLBF, 0);
	val = 0;
	if (optind < argc)
	{
		while (optind < argc)
			printnames (argv[optind++], family, aflags, nflags, single);
	}
	else
		val = printnamesf (stdin, family, aflags, nflags, single);

	return val;
}
